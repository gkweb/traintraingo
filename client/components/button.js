import styled from 'styled-components'

export const ButtonElem = styled.button `
  display: block;
  box-sizing: border-box;
  background-color: ${props => props.theme.secondaryBg};
  border: 1px solid ${props => props.theme.secondaryBg};
  color: ${props => props.theme.primary};

  &:focus {
    outline: 0;
    box-shadow: 0 0 ${props => props.theme.primary} 2px;
  }

  &:hover,
  &:focus {
    background-color: ${props => props.theme.tertiaryBg};
  }
`
