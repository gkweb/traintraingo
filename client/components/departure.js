import styled from 'styled-components'
import moment from 'moment'
import Disruption from './disruption'
import RouteStops from './route-stops'
import GoodService from './good-service'
import PropTypes from 'prop-types'

const ContainerElem = styled.div`
  padding: 1.5em 1em 1em;
  width: 100%;
  border: 1px solid ${props => props.theme.tertiaryBg};
  background: ${props => props.theme.inversePrimaryBg};
  color: ${props => props.theme.inversePrimary};
  box-shadow: -0.0625em 0.125em 0.5em -0.125em rgba(0, 0, 0, 0.125);
`

const TitleElem = styled.h2`
  margin: 0 0 0.75em;
`

const InfoElem = styled.p`
  margin: 0;

  &:not(:last-child) {
    margin: 0 0 1em;
  }
`

const DepartingTimeElem = styled.p`
  font-size: 1.5em;
  margin: 0 0 0.75em;
`

const PlatformElem = styled.p`
  font-size: 1.125em;
  margin: 0;

  &:not(:last-child) {
    margin: 0 0 1em;
  }
`

const Departure = ({
  lineName,
  directionName,
  platformNumber,
  scheduledDep,
  estimatedDep,
  disruptions,
  runId,
}) => {
  const scheduled = moment(scheduledDep).format('HH:mm')
  const estimated = estimatedDep ? moment(estimatedDep).fromNow() : '-' // Estimates into the future are null

  return (
    <ContainerElem>
      <TitleElem>
        {scheduled} to {directionName}
      </TitleElem>
      {lineName ? <InfoElem>Line: {lineName}</InfoElem> : null}
      <DepartingTimeElem>{estimated}</DepartingTimeElem>
      <PlatformElem>
        Platform: {platformNumber ? platformNumber : '-'}
      </PlatformElem>
      <RouteStops runId={runId} />
      {disruptions && disruptions.length > 0 ? (
        <Disruption disruptionData={disruptions} />
      ) : (
        <GoodService />
      )}
    </ContainerElem>
  )
}

Departure.propTypes = {
  lineName: PropTypes.string,
  directionName: PropTypes.string,
  platformNumber: PropTypes.number,
  scheduledDep: PropTypes.string,
  estimatedDep: PropTypes.string,
  disruptions: PropTypes.array,
  runId: PropTypes.number,
}

export default Departure
