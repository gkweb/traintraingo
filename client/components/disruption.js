import React, {Component} from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import {Chevron} from './icon'

const ContainerElem = styled.div`
  display: flex;
  flex-flow: column nowrap;
`

const ChevronIconElem = styled(Chevron)`
  max-width: 1em;
  margin-left: auto;
  transform: ${props => (props.isOpen ? 'rotate(-90deg)' : 'rotate(0)')}
`

const NoticeElem = styled.svg`
  display: block;
  width: 1.25em;
  height: 1.25em;
  margin-right: .25em;
  fill: ${props => props.theme.highlightSecondary};
`

const TitleElem = styled.button`
  display: flex;
  padding: .5em;
  width: 100%;
  border: 0;
  margin: 0;
  font-size: 1em;
  font-weight: bold;
  font-family:'Lustria', serif;
  color: ${props => props.theme.inversePrimary};
  background: none;
  position: relative;
`

const TitleTextElem = styled.p`
  margin: 0;
`

const ContentElem = styled.div`
  padding: 1em;
  background-color: ${props => props.theme.primaryBg};
  color: ${props => props.theme.primary};
`

const DisruptionListElem = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`
const DisruptionItemElem = styled.li`
  margin-bottom: 1em;
`

const DisruptionTitleElem = styled.span`
  display: block;
  margin-bottom: .5em;
  font-weight: bold;
`

const DisruptionDescElem = styled.span`
  display: block;
  margin-bottom: .5em;
`

const DisruptionReadMoreElem = styled.a`
  display: block;
  color: ${props => props.theme.link};
`

class Disruption extends React.Component {
  constructor() {
    super()
    this.state = {
      visible: false
    }

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick (event) {
    // Toggle
    this.setState({visible: !this.state.visible})
  }

  render() {
    let d = null
    let content = null

    if (this.props.disruptionData && this.props.disruptionData.length > 0) {
      if (this.state.visible) {
        content = <ContentElem>
          <DisruptionListElem>
          {this.props.disruptionData.map((val, index) => (
            <DisruptionItemElem key={index}>
              <DisruptionTitleElem>{val.title}</DisruptionTitleElem>
              <DisruptionDescElem>{val.description}
                {val.url ? <DisruptionReadMoreElem href={val.url} colour={val.colour}>Read more</DisruptionReadMoreElem> : null}
              </DisruptionDescElem>
            </DisruptionItemElem>
            )
          )}
          </DisruptionListElem>
        </ContentElem>
      }

      d = <ContainerElem>
      <TitleElem onClick={this.handleClick}>
        <NoticeElem viewBox="0 0 32 32">
          <path d="M15.5 3c-7.456 0-13.5 6.044-13.5 13.5s6.044 13.5 13.5 13.5 13.5-6.044 13.5-13.5-6.044-13.5-13.5-13.5zM15.5 27c-5.799 0-10.5-4.701-10.5-10.5s4.701-10.5 10.5-10.5 10.5 4.701 10.5 10.5-4.701 10.5-10.5 10.5zM15.5 10c-0.828 0-1.5 0.671-1.5 1.5v5.062c0 0.828 0.672 1.5 1.5 1.5s1.5-0.672 1.5-1.5v-5.062c0-0.829-0.672-1.5-1.5-1.5zM15.5 20c-0.828 0-1.5 0.672-1.5 1.5s0.672 1.5 1.5 1.5 1.5-0.672 1.5-1.5-0.672-1.5-1.5-1.5z"></path>
        </NoticeElem>
        <TitleTextElem>Possible disruption</TitleTextElem>
        <ChevronIconElem isOpen={this.state.visible}/>
      </TitleElem>
      {content}
    </ContainerElem>
    }

    return d
  }
}

Disruption.propTypes = {
  disruptionData: PropTypes.array
}

export default Disruption
