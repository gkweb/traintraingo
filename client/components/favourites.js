import React, {Component} from 'react'
import styled from 'styled-components'
import Link from 'next/link'
import {H3} from './text'
import {getItems} from './../lib/local-storage'
import {Bookmark} from './icon'

const FavoriteListElem = styled.ul`
  display: flex;
  flex-flow: row wrap;
  margin: 0 0 0 -.5em;
  padding: 0;
`

const FavoriteItemElem = styled.li`
  display: inline-flex;
  margin: .5em;
`

const FavoriteLinkElem = styled.a`
  display: flex;
  align-items: center;
  width: 100%;
  max-width: 30em;
  padding: 1em;
  border: 1px solid black;
  background: ${props => props.theme.primary};
  box-shadow: -.0625em .125em .5em -.125em rgba(0,0,0,.7);
  color: ${props => props.theme.tertiary};
  text-decoration: none;
  transition: transform .125s ease;

  &:hover,
  &:focus {
    transform: scale(1.05);
  }

  span:last-child {
    margin-left: .5em;
    color: ${props => props.theme.highlightPrimary};
  }
`

const FavouritesContainerElem = styled.section`
  padding: 1em 1em 4em;
  background-color: #fff;

  @media screen and (min-width: 50em) {
    padding: 1em 1em 4em;
    margin-top: 3em;
  }
`

class Favourites extends React.Component {
  constructor() {
    super()
    this.state = {
      favourites: {}
    }
  }

  componentDidMount() {
    const favourites = getItems()
    if (favourites) this.setState({favourites: favourites})
  }

  render() {
    let r

    if (this.state.favourites && Object.keys(this.state.favourites).length > 0) {
      r = <FavouritesContainerElem>
      <H3>Favourites</H3>
      <FavoriteListElem>
        {Object.keys(this.state.favourites).map((val, ind) => (
          <FavoriteItemElem key={ind}>
            <Link href={`/departures/${val}`} as={`/departures/${val}`} passHref>
              <FavoriteLinkElem>
                <span>{this.state.favourites[val]}</span>
                <span><Bookmark/></span>
              </FavoriteLinkElem>
            </Link>
        </FavoriteItemElem>
        ))}
      </FavoriteListElem>
      </FavouritesContainerElem>
    } else {
      r = null
    }

    return r
  }
}

export default Favourites
