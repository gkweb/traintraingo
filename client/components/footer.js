import styled from 'styled-components'
import Link from 'next/link'

const FooterElem = styled.footer`
  padding: 0 1em 3em;
  margin-top: 5em;
  color: ${props => props.theme.secondary};
`

const LinkElem = styled.a`
  color: ${props => props.theme.secondary};
  &:visited {
    color: ${props => props.theme.secondary};
  }

  &:hover,
  &:focus {
    text-decoration: underline;
  }
`

const Footer = () => (
  <FooterElem>
    <p>
      Data supplied by ptv (Public Transport Victoria). Obtain your API Access{' '}
      <LinkElem href="https://www.data.vic.gov.au/data/dataset/ptv-timetable-api">
        here
      </LinkElem>
    </p>
    <LinkElem href="https://gladekettle.com.au" target="_blank">
      Created by gkweb
    </LinkElem>
  </FooterElem>
)

export default Footer
