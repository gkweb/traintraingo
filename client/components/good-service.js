import styled from 'styled-components'
import { TickIcon } from './icon'

const Container = styled.div`
  display: flex;
  flex-flow: row nowrap;
  padding: 0.5em;
`

const Icon = styled(TickIcon)`
  width: 1.25em;
  height: 1.25em;
  margin-right: 0.25em;
  fill: ${props => props.theme.highlightTertiary};
  color: ${props => props.theme.highlightTertiary};
`

const TextBlock = styled.span``

const GoodService = () => (
  <Container>
    <Icon />
    <TextBlock>Good Service</TextBlock>
  </Container>
)

export default GoodService
