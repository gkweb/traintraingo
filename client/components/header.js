import styled from 'styled-components'
import PropTypes from 'prop-types'
import Logo from './../components/logo-no-text'
import Save from './../components/save'
import Link from 'next/link'
import { updateItem, removeItem, getItem } from './../lib/local-storage'

const HeaderElem = styled.header`
  display: flex;
  flex-wrap: row wrap;
  align-items: center;
  padding: 2em 1em;
  width: 100%;
  height: 4em;
  border-bottom: 1px solid rgba(0, 0, 0, .25);
  background-color: ${props => props.theme.primaryBg};
  margin-bottom: 1em;
  position: fixed;
  top: 0;
  left: 0;
  z-index: 200;

  @media screen and (min-width: 50em) {
    height: 6.625em;
  }
`

const TitleElem = styled.h1`
  margin: 0 0 0 .5em;
  color: ${props => props.theme.primary};
  font-size: 1.25em;

  @media screen and (min-width: 50em) {
    font-size: 2em;
  }
`

const LogoContainer = styled.div`
  max-width: 4em;
  margin: 0;

  @media screen and (max-width: 50em) {
    max-width: 2em;
  }
`

class Header extends React.Component {
  constructor() {
    super()
    this.state = {
      isSaved: false,
      clientReady: false
    }

    this.handleSave = this.handleSave.bind(this)
  }

  handleSave(event) {
    this.setState({
      isSaved: !this.state.isSaved
    })

    if (this.state.isSaved) {
      // remove
      removeItem(this.props.stopId)
    } else {
      // save
      updateItem(this.props.stopId, this.props.headerTitle)
    }
  }

  componentDidMount() {
    const isSaved = (getItem(this.props.stopId))

    this.setState({
      isSaved: isSaved,
      clientReady: true
    })
  }

  render() {
    const {
      headerTitle,
      isLoading
    } = this.props

    return (
      <HeaderElem>
      <Link href={'/'} passHref prefetch>
        <a>
          <LogoContainer>
            <Logo />
          </LogoContainer>
        </a>
      </Link>
      {(headerTitle && !isLoading) ? <TitleElem>{headerTitle}</TitleElem> : null}
      {!isLoading ? <Save isSaved={this.state.isSaved} handler={this.handleSave} /> : null }
    </HeaderElem>
    )
  }
}

Header.propTypes = {
  headerTitle: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired
}

export default Header
