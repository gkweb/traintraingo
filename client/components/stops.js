import styled from 'styled-components'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import Link from 'next/link'

const StopsLoadingTextElem = styled.p`
  color: ${props => props.theme.primary};
`

const StopButtonElem = styled.a`
  display: block;
  box-sizing: border-box;
  width: 100%;
  padding: 1em;
  background-color: ${props => props.theme.linkBg};
  color: ${props => props.theme.link};
  text-decoration: none;

  &:hover,
  &:focus {
    text-decoration: underline;
    background-color: ${props => props.theme.linkHoverBg};
  }
`

const StopContainerElem = styled.div`
  padding-bottom: 2em;
  margin-top: 1em;
  color: ${props => props.theme.primary};
`

const StopResultsTitle = styled.h3`
  margin-bottom: 1rem;
`

const Stops = ({ search_term }) => {
  return (
    <Query query={STOPS_QUERY} variables={{ search_term }}>
      {({ loading, error, data }) => {
        if (error) return <p>Error :(</p>
        if (loading && (search_term && search_term.length >= 2))
          return <StopsLoadingTextElem>Loading...</StopsLoadingTextElem>
        if (!data.stops) return null
        if (data.stops && data.stops.length <= 0) return null

        let v = null
        if (
          data.stops &&
          ((data.stops || []).length > 0 && typeof data.stops === 'object')
        ) {
          v = (
            <StopContainerElem>
              <StopResultsTitle>Results:</StopResultsTitle>
              {data.stops.map((stop, index) => (
                <Link
                  key={stop.stop_id}
                  href={`/departures/?stop_id=${stop.stop_id}`}
                  as={`/departures/${stop.stop_id}`}
                  passHref
                >
                  <StopButtonElem>{stop.stop_name}</StopButtonElem>
                </Link>
              ))}
            </StopContainerElem>
          )
        } else {
          v = <div>Nothing found</div>
        }
        return v
      }}
    </Query>
  )
}

const STOPS_QUERY = gql`
  query($search_term: String!) {
    stops(search_term: $search_term) {
      stop_id
      stop_name
    }
  }
`
export { Stops, STOPS_QUERY }
