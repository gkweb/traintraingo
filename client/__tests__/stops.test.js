import React from 'react'
import 'jest-styled-components'
import { shallow } from 'enzyme'
import { Stops, STOPS_QUERY } from './../components/stops'
import { MockedProvider } from 'react-apollo/test-utils'
import waitForExpect from 'wait-for-expect'
import renderer from 'react-test-renderer'

// Stops Mock
const mocks = [
  {
    request: {
      query: STOPS_QUERY,
      variables: {
        search_term: 'Flagstaff',
      },
    },
    result: {
      data: {
        stops: [
          {
            stop_id: '1068',
            stop_name: 'Flagstaff Station',
          },
        ],
      },
    },
  },
]

test('User can search and station name suggestion link is displayed', async () => {
  const c = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Stops search_term={'Flagstaff'} />
    </MockedProvider>
  )

  await waitForExpect(() => {
    expect(JSON.stringify(c.toJSON())).toMatch('Flagstaff Station')
  })
})

it('Should render loading state with empty result', () => {
  const component = renderer.create(
    <MockedProvider mocks={[]}>
      <Stops search_term={'Flagstaff'} />
    </MockedProvider>
  )

  expect(JSON.stringify(component.toJSON())).toMatch('Loading...')
})

test('Home / Index Snapshot', () => {
  const c = shallow(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Stops search_term={'Flagstaff'} />
    </MockedProvider>
  )
  expect(c).toMatchSnapshot()
})
