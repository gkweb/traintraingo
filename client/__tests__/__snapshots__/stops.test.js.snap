// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home / Index Snapshot 1`] = `
<ApolloProvider
  client={
    ApolloClient {
      "cache": InMemoryCache {
        "addTypename": false,
        "cacheKeyRoot": CacheKeyNode {
          "children": null,
          "key": null,
        },
        "config": Object {
          "addTypename": false,
          "dataIdFromObject": [Function],
          "fragmentMatcher": HeuristicFragmentMatcher {},
          "resultCaching": true,
        },
        "data": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "maybeBroadcastWatch": [Function],
        "optimisticData": DepTrackingCache {
          "data": Object {},
          "depend": [Function],
        },
        "silenceBroadcast": false,
        "storeReader": StoreReader {
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "executeSelectionSet": [Function],
          "executeStoreQuery": [Function],
        },
        "storeWriter": StoreWriter {},
        "typenameDocumentCache": Map {},
        "watches": Set {},
      },
      "clearStoreCallbacks": Array [],
      "clientAwareness": Object {},
      "defaultOptions": Object {},
      "disableNetworkFetches": false,
      "link": ApolloLink {
        "request": [Function],
      },
      "localState": LocalState {
        "cache": InMemoryCache {
          "addTypename": false,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": false,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
        "client": [Circular],
        "resolvers": Object {},
      },
      "mutate": [Function],
      "query": [Function],
      "queryDeduplication": true,
      "reFetchObservableQueries": [Function],
      "resetStore": [Function],
      "resetStoreCallbacks": Array [],
      "ssrMode": false,
      "store": DataStore {
        "cache": InMemoryCache {
          "addTypename": false,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": false,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
      },
      "typeDefs": undefined,
      "version": "2.5.1",
      "watchQuery": [Function],
    }
  }
>
  <Stops
    search_term="Flagstaff"
  />
</ApolloProvider>
`;
